// xmind-generator/lib/cjs/internal/builder/sheet-builder.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeSheetBuilder = makeSheetBuilder;
var _sheet = require("../model/sheet.cjs");
var _types = require("./types.cjs");
function makeSheetBuilder(title) {
  let rootTopicBuilder;
  const relationshipBuilders = [];
  return {
    rootTopic(topicBuilder) {
      rootTopicBuilder = topicBuilder;
      return this;
    },
    relationships(relationships) {
      relationshipBuilders.push(...relationships);
      return this;
    },
    title(sheetTitle) {
      title = sheetTitle;
      return this;
    },
    build() {
      const {
        topic: rootTopic,
        reference
      } = (0, _types.asBuilder)(rootTopicBuilder)?.build() ?? {};
      const sheet = new _sheet.Sheet(title, rootTopic);
      relationshipBuilders.forEach(builder => {
        const {
          title: title2,
          from,
          to
        } = (0, _types.asBuilder)(builder).build();
        sheet.addRelationship(title2, reference.fetch(from).id, reference.fetch(to).id);
      });
      return sheet;
    }
  };
}
