// xmind-generator/lib/cjs/internal/model/sheet.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Sheet = void 0;
var _common = require("../common.cjs");
var _topic = require("./topic.cjs");
var _relationship = require("./relationship.cjs");
class Sheet {
  title;
  id;
  _rootTopic;
  _relationships;
  constructor(title, rootTopic) {
    this.id = (0, _common.uuid)();
    this.title = title ?? "";
    this._rootTopic = rootTopic ?? null;
    this._relationships = [];
  }
  get rootTopic() {
    return this._rootTopic;
  }
  get relationships() {
    return this._relationships;
  }
  query(topicId) {
    return this._rootTopic?.query(topicId) ?? null;
  }
  addRootTopic(title) {
    if (this._rootTopic) {
      throw new Error("Duplicated root topic creation");
    }
    this._rootTopic = new _topic.Topic(title);
    return this._rootTopic;
  }
  removeRootTopic() {
    this._rootTopic = null;
  }
  addRelationship(title, from, to) {
    const relationship = new _relationship.Relationship(title, from, to);
    this._relationships.push(relationship);
    return relationship;
  }
}
exports.Sheet = Sheet;
